✨amp-jwplayer: Add support for contextual article matching

boyntoni opened the pull request:
Implements additional contextual article matching attributes for amp-jwplayer component

estherkim committed:
Gulp serve before gulp e2e (#20424)

dparikh committed:
SwG release 0.1.22.42 (#20482)

Commit:
✅Amp-skimlinks end to end tests (#20211)

Add end to end tests fo amp-skimlinks extension.

#### Notes:
In order to test amp-skimlinks script, we introduced a new **optional** JSON config to overwrite the request endpoints used by amp-skimlinks. The config (used only for the end to end tests) can be inserted inside the `<amp-skimlinks>` tag as follow:

```html
<amp-skimlinks
    layout="nodisplay"
    publisher-code="123X123"
    tracking="true"
>
    <script type="application/json">
    {
        "pageTrackingUrl": "${pageTrackingUrl}",
        "linksTrackingUrl": "${linksTrackingUrl}",
        "nonAffiliateTrackingUrl": "${nonAffiliateTrackingUrl}",
        "waypointUrl": "${waypointUrl}"
    }
    </script>
</amp-skimlinks>
```

This allows us to overwrite the API endpoints to use the "request-bank" instead, so we can later validate if an API call has happened as expected or not.

cvializ committed:
Document the attributes and values of amp-date-display (#20483)

jridgewell committed:
Add RC binary type (#20463)

We're releasing an RC build, which will be 100% the same as the code
that _wil be released_ to production next week. The canary type sticks
around, but it is (as it has always been) an experimental configuration
of next week's code.

Proper experiment comparisons should be done `canary <-> rc`, and `rc
<-> control`. Comparing `canary <-> control` will not give the correct
experiment configurations, because it tests experimental code _and_
experimental configurations.

alabiaga committed:
remove old sauce lab credit snippet (#20486)

thanks for noticing Esther!

Commit:
✅Add carousel v2 tests. (#20469)

- Add scrollBy method for functional tests.
- Add tests for mixed-length carousels.
- Add tests for responsive attributes.
- Add tests for max-swipe configuration.

dYale commented on extensions/amp-jwplayer/0.1/amp-jwplayer.js:
```diff
@@ -53,25 +65,39 @@ class AmpJWPlayer extends AMP.BaseElement {
   /** @override */
   layoutCallback() {
     const iframe = this.element.ownerDocument.createElement('iframe');
+    const contentSearch = this.contentSearch_ ? 'search=' + encodeURIComponent(this.contentSearch_) : ''
```
Do we want to encode these individually? Or can we just do at the end with the other encodings? Line 99

estherkim committed:
🏗 Move e2e modules to its own package.json (#20487)

danielrozenberg committed:
🏗🚮 Remove legacy bundle-size check (#19541)

* Remove legacy bundle-size check

* Remove unnecessary space

alanorozco committed:
🐛 Always reapply docking target on resize (#20491)

jpettitt committed:
enable pre-render mode (#20490)

Commit:
Revert new user error in purifier. (#20501)

j343my committed:
fix(doc): fix adyoulike readme typo error (#20475)

TcsEmotion committed:
Added TcsEmotion ad support to amp-ad (#20367)

* added new ad network tcsemotion

* amp-ad network for tcsemotion

* amp-ad network for tcsemotion

* amp-ad network for tcsemotion

* deleting package-lock.json

deleting package-lock,json, because this change not required in PR

* Update tcsemotion.js

* Update tcsemotion.js

* Update tcsemotion.md

* Update tcsemotion.md

* new lines removed

* new line removed

* Update tcsemotion.md

torch2424 committed:
Fixed OnNetwork amp ad example (#20507)

rsimha committed:
🏗 Fetch and print sauce labs status before connecting (#20504)

cvializ committed:
🚮 Remove amp-date-display experiment (#20505)

* Remove amp-date-display experiment

* Update documentation

calebcordry committed:
🐛story: fix fie querySelector (#20506)

* fix querySelector error

* use element

alanorozco committed:
💅 Docking and undocking transition polish (#20508)

- `Dock in two steps` is now `multiple steps`. I haven't figured out what causes (multiple) browsers to visually glitch if transitioning in one step. It looks like it's related to GPU layer transfer but I haven't looked fully into it. The visual glitch prevented is reproducible in Chrome and Safari.

- `TargetAreaDef` is gone in favor of standard `LayoutRectDef`s.

- Enable controls after docking transition so they won't appear if the cursor is not hovering the final bounding box.

- Pause video when undocking if it's out of view.

- Don't transition video if it's out of view.

- Properly calculate undocking transition final bounds `Y`. (Fixes #20215)

- Set min-width and min-height to prevent relative sizing on cramped viewports.

- Multiple cleanups.

cathyxz committed:
[amp-list] guarantee space for load-more button instead of overlaying on content (#20332)

* Add draft fix for amp-list load-more overflow

* Undo force element to change size

* Adapt margin bottom to size of load-mmore button

* Unbreak unit tests

* Took load-more elements out of container

* Mark as ok

* Rename initialization functions for clarity

* Change div to span to unbreak integration tests

* Rework load-more-element height implementation to remove applyFillContent

* Move adjustment for load-more elements to layout callback instead of build callback

* Use correct value of viewport top and height

* Add comment clarifying conditional removal of applyFillContent

* Change viewport.onScroll to viewport.onChanged

* Only load more if content height has not reached end of list

* Remove unnecessary debounce and don't load more items if overflow active

sgorelkin committed:
Add rbinfox (#20451)

* Initial version

* fix unit test problems

* fix unit test

* fix unit test

* just for run percy test

* rollback new line

* just run test

alanorozco committed:
✨Notify amp-lightbox-gallery of new lightbox elements (#20485)

* Notify amp-lightbox-gallery of new lightbox elements

* small perf improvement

* add manual test

* dep-check

* Ignore visited items in explicit check

* oops

* tests, do not use specific event

* remove unused

* lint

* remove irrelevant change

* types

* multiple item manual test

* address review

* log

* Use use-case specific event

* lint

* dep check

* types

* move constant in frustration

* typo

* oops

* compacc

chenzl3000 committed:
Move preventDefault before amp-ad-exit filters (#20243)

* Fixed that border protection / click delay is not working on <a> element
   because preventDefault is not called when filter return false.

Commit:
Avoid clipping bottom of carousel slides on desktop. (#20513)

zhouyx committed:
Apply length check to consent string (#20493)

* lstring engtth check

* minor change

torch2424 committed:
🐛 amp-recaptcha-input: Specified / Check bootstrap frame origin in the recaptcha service (#20358)

* Added origin checks for the recaptcha iframe

* Added tests, fixed precommit

* Made PR Comments

* Made PR Comments

alanorozco committed:
🐛 Hide gallery button when displaying single item (#20516)

alanorozco committed:
🐛 [auto-lightbox] Relax schema parsing (#20518)

zombifier committed:
Clean up inabox-rov experiment. (#20111)

The checks were removed in #18606.

gmajoulet committed:
Pause the story when the landscape overlay warning is displayed. (#20498)

nashdb committed:
✨Adding Click Target to Amp-ad-exit (#20373)

* Add config g for properties and clickTarget Property.  Currently untested

* Renaming properties to become behaviors

* Fixing small formatting errors

* Fixing more small formatting errors

* Refactoring clickTarget if-statement

* Adding comment defining the clickTarget behavior

* Removing BehaviorsDef and adding unit test

* Small Formatting Errors fix

* Updating amp-ad-exit documentation.  Renaming behavior -> behaviors.  Adding fix for broken test

* Removing unnecessary parens

camel2243 committed:
✨ Improve kuadio ad network optional params data-ref (#19813)

* Support optional params data-ref

* Replace window.context.canonicalUrl with global.context.canonicalUrl

* Fix lint

tockn committed:
✨ Support new ad type for Zucks ad network (#20206)

* support zoe (only display)

update ads/zucks.md

update ads.amp.html to display zoe

review修正

* global.window -> global

newmuis committed:
📖 Add documentation for localizing AMP stories (#20503)

* Update amp-story.md

* Add space before lang attribute

gmajoulet committed:
Story landscape experience and fullbleed desktop documentation. (#20421)

* Landscape experience and fullbleed desktop documentation.

* s/amp-story-page/amp-story/

* Punctuation and updated gifs to `amp-anim`

gmajoulet committed:
Fix an image link typo in the amp-story documentation. (#20532)

Fix an image link typo in the amp-story documentation.

zombifier committed:
Reenable some inabox tests on Safari (#20381)

* Reenable some inabox tests on Safari

* verify that the host script is loaded

* Make loading the host script work during testing.

peterjosling committed:
✨ amp-next-page: Send document title/URL changes to the viewer (#20484)

* Add support for url and canonicalUrl parameters in History

* Use History service to update page title and path when switching documents

Will send messages through the viewer integration API.

* add type

* Update history-impl.js

* use describes for tests.

* make tests side-effect free

alabiaga committed:
launch amp-list-viewport-resize to 75% in prod (#20552)

#20309

aghassemi committed:
🐛 Fix amp-list documentation (#20553)

This is breaking validation since it ends up as `<a> <amp-list> </a>`

torch2424 committed:
♻️ amp-recaptcha-input: Allow passing the recaptcha API url from the frame (#20539)

* Started the integration test and mess around with bootstrap frame

* Left a note for monday

* Made tons of progress on the integration test

* Got the text replace working with the mock file

Need to write recaptcha.mock.js

* TIL Integration test watching wont build your extension.

* Finished integration test, fixed precommit errors

* Split up PRs

* Forgot a file to split out

* Made requested changes

alanorozco committed:
🗑️ Remove `video-dock` experiment guard. (#20413)

mrjoro committed:
provide more details on design reviews (#20533)

mrjoro committed:
add avenues for CLA help (#20521)

cvializ committed:
✨Add AMP layer to E2E tests (#20502)

* Add AmpDriver for toggleExperiment. Enable cli debug flags

* Fix rebase munge

* Add AmpDriver to new tests and mocha setup

* Review feedback

* Use the document under test to set the experiment vs the root page

* Use spread argument

* Fix eval lint

gmajoulet committed:
Fix closing the page attachment through popping the history state. (#20443)

erwinmombay committed:
temporarily disable saucelabs because of flakiness (#20566)

cvializ committed:
✅Fix amp-date-display flakes (#20568)

* Use fake timer to avoid test flakes

* Reference all Date constructors from element window reference

alanorozco committed:
✅Add overflow prevention test (#20546)

alanorozco committed:
✅Validate `dock` for `amp-brightcove` (#20567)

alanorozco committed:
✅Validate `dock` for `amp-youtube` (#20554)

alanorozco committed:
📖 Update `amp-video-docking` CSS reference (#20571)

zhangsu committed:
Fix amp-sidebar keyboard event forwarding (#20557)

It turns out that the close_() function may be a no-op if the sidebar
has already been closed, in which case we should not prevent the
default.

Currently, just by having an amp-sidebar in the AMP document, and
clicking on a blank region in the document would prevent the default and
stop forwarding the key event to the viewer. This is because the
keyboard event listener is attached onto the `document` object. We still
need to forward the keyboard event when the sidebar is already closed.

alanorozco committed:
🐛Reparent placeholder to account for incorrect positioning (#20562)

alanorozco committed:
✨Get placeholder background from `placeholder` srcs (#20563)

To account for the entire poster image API surface and cases where placeholder is dynamically set, e.g. `amp-youtube`.

Commit:
✅ Additional carousel v2 end to end tests (#20522)

- Reduce minimum auto advance interval to 1 second, matching carousel v1
- Support amp-bind for carousel v2
  * Still need to add validation rules
- Add tests for auto advance
- Add tests for slide grouping
- Add tests for a vertical carousel
- Add tests for non-looping carousels
- Add test for amp-bind with responsive attributes

Commit:
Changes to make carousel v2 work with lightbox gallery. (#20558)

- Change lightbox gallery selectors to also look for carousel v2 slides.
- Fire an event from the carousel internals for `<amp-carousel>` to listen to.
- Fire the `slideChanged` event from carousel v2.
- Implement `showSlideWhenReady`.
- Add height: 100% for slides so that the image viewer sizes correctly.
- Add `object-fit: contain` to match carousel v1.

bramanudom committed:
📖Add sidebar in stories documentation (#20385)

* added documentation for sidebar in stories in the main sidebar doc

* fix typos, add xreference for amp-story doc

* remove extra line

* fixed some typos and wording changes

* implemented requested changes

* Small punctuation nits

* made requested changes

* removed incorrect bullet

* add more detail to cross-reference

alabiaga committed:
Fix amp-form server side rendering (#20511)

b/123310281

Commit:
Clean up bindings after amp-list reset. (#20570)

cvializ committed:
🐛Fix amp-inputmask attributes stripping in amp-mustache 0.2.  (#20515)

* Fix amp-inputmask attributes stripping in amp-mustache 0.2. Update example

* Only add the new attribute mask-output

* Add test

rsimha committed:
🏗 Show progress of `browserify` transform prior to running tests (#20573)

jridgewell committed:
Enables layers in canary (#20512)

gmajoulet committed:
Fix toast z-index so it appears above the share menu. (#20565)

alabiaga committed:
wrap update font size in mutateElement (#20585)

alanorozco committed:
✅ Improve `amp-video-docking` unit test coverage (#20583)

- Restructure `describes.repeated` so that only tests that discern between slot target and corner target form part of the block.

- Group tests for `getTargetFor_` and `updateOnPositionChange_`.

- Unskip tests that were skipped only for slot.

- Add tests for:
    - `placeAt_`
    - `dockInTransferLayerStep_`
    - `getTargetArea_`
    - `getTargetAreaFromPos_`
    - `getTargetAreaFromSlot_`
    - `getDims_`
    - `dock_`
    - `setCurrentlyDocked_`
    - `undock_`
    - `trigger_`

jsalgueiro committed:
✨ Integrate AdSensor network into AMP (#20375)

* Integrate AdSensor network into AMP

* Fix Copyright year

* Fix Copyright year

* Fix typo

* removed renderStartImplemented from the ads _config

* missing trailing comma

* WSD-1287:  reverted implement renderStart to help serve ads fast

Commit:
Use smaller images for carousel v2 tests. (#20579)

torch2424 committed:
🐛Moved toggle order for Consent Iframe (#20588)

* Made a fix

* Used a trey catch, window.name will always be a string

cvializ committed:
🐛Prevent letters in credit cards. Allow editing beginning. (#20576)

* Prevent letters in credit cards. Allow editing beginning.

* Simplify boolean logic

aghassemi committed:
🐛 amp-img: make intrinsic sizer image hidden to ATs (#20581)

* 🐛 amp-img: make intrinsic sizer image hidden to ATs

Fixes https://github.com/ampproject/amphtml/issues/20578

* lint

calebcordry committed:
✨Pass doc into validator ui using url (#20391)

* read param

* use fragment

* set default to amp

calebcordry committed:
✨Allow iframe ping in configRewriter (#20561)

* allow iframe

* transport => ping

* add server to whitelist

twifkak committed:
Validator rollup (#20597)

* cl/230559764 Small validator refactoring - move the amp-custom byte size adjustment down into the validator Match call.

* cl/230603165 add amp-payment-google-inline-async rules and tests

* cl/230963421 Provide a validator option to subtract the length of URLs from the CSS cdata strings, when counting against validation lengths.

* cl/231000057 Make valid tests valid for amp-payment-google-inline-async

* cl/231319523 github commit msg missing or malformed

* cl/231332520 Revision bump for #20567

* cl/231332548 Revision bump for #20554

* cl/231613739 Fix lint issues for validator j/s files

newmuis committed:
📖 Add a README for contributor-facing documentation on adding strings/translations for stories (#20510)

* Create README.md

* Update README.md

bramanudom committed:
✨ Add Analytics for branching (#20496)

* WIP adding advacementMode and previousPageId as analytics variables

* WIP: made more changes following the onNavigationState update pattern instead

* moved around some code, got rid of excess

* added tests

* tested with helloworld

* cleaned code

* linted, cleaned more

* got rid of changes to examples, removed unecessary changes

* fixed typo, oops!

* removed analytics trigger, addressed nits

* removed analytics test, added variable service test

* removed advance_to as an advancement mode, capture other instances of manual advancing

* addressed requested changes

* ?string shorthand instead of string|null

* fixed navigation state tests

zhouyx committed:
CMP Integration Doc (#20396)

* draft doc

* some changes

* acomments

* address comments

* Small nit

torch2424 committed:
Revert "🐛async-input: Fixed NonXHR GET on amp-form with Async Input Elements (#20362)" (#20601)

This reverts commit de0ea8e2bdd449181602bbb33c4d114d105c7afa.

Commit:
Support pattern validation on textarea (#20544)

* Support pattern validation on textarea.

* Fix getValidationFor().

* Fix tests, add textarea validity tests.

* Test custom validations for textarea. Make regex multi-line.

* Validation rules for pattern and pattern-error.

* Actually, we don't need 'pattern-error'.

* Use devAssert.

* Don't override verify errors; documentation.

* Add unit test for error overriding case.

sohanirao committed:
SwG Release 0.1.22.43 (#20606)

estherkim committed:
Update gulp e2e task (#20605)

calebcordry committed:
Document new `doc=` feature for validator UI. (#20610)

* docs

* change link

Commit:
Force the subscriptions dialog into the fixed layer to avoid collisions with border (#20608)

calebcordry committed:
more choices (#20614)

alanorozco committed:
🚀 [devdash] Render file list statically as placeholder (#20600)

Prevents flash of unloaded content!

Also:

- automatically finds required AMP extensions in document string.
- fixes a validation error for `amp-mustache`.

Closes #19896.

AramZS committed:
Proposal to add Learn More as a CTA option for ads (#20619)

We have had an advertiser interested in using this specific CTA at The Washington Post.

gmajoulet committed:
Page attachment call to action animation and design review. (#20612)

* Page attachment call to action animation and design review.

* Restarting the animation every time the page becomes active.

cathyxz committed:
Add origin experiments check for amp-list (#20592)

* Check for origin trial experiments in amp-list

* Fix type check on origin-experiments-impl

* Whitelist origin experiments dependency check

* Refactor load more creation and show/hide logic out of amp-list

* Fix incorrect variable

* Stub getAmpDoc on the AmpDocService prototype

* Short circuit origin trial logic if no amp-list-load-more attribute

Commit:
amp-script: Fetch locally, add size limit (#20427)

* Fetch scripts in AMP land instead of in worker-dom.

* Add size cap.

* Upgrade worker-dom to 0.2.9.

* Update yarn.lock.

* Require element uniqueness until global size limit.

* Upgrade to worker-dom 0.2.10 so it downgrades to dompurify 1.0.8. :D

* Remove babel-regenerator-runtime.

estherkim committed:
🏗 Add e2e section to TESTING.md (#20621)

Enriqe committed:
Unify store actions for embedded components and place tooltip on touch target (#20430)

* unify store actions for components and place tooltip on touch target

* fix desktop

* types

* entering animation duration

* always updates store on state change, introduces compare fn in store state property, moves states to store.

* types store service

* dont hide system layer on desktop panels view, change oncomponentupdate impl, reviews

taymonbeal committed:
Create ads/google/a4a/shared/ directory (#20620)

This is intended to make <amp-ad type=adsense> and
<amp-ad type=doubleclick> code easier to maintain going forward by
keeping it in sync with proprietary Google code located elsewhere that
hits the same endpoints.

cvializ committed:
🐛Remove mask behavior if user deletes from middle of date (#20613)

* Remove mask behavior if user deletes from middle of date

* Add patch file.

zikas committed:
🐛 Use analytics macros to expand varGroups. (#20603)

* Use analytics macros to expand varGroups.

* Modify tests to make sure getMacros() is called.

* Lint cleanup.

* Lint cleanup.

* Use analytics macros to expand varGroups.

* Modify tests to make sure getMacros() is called.

* Lint cleanup.

* Lint cleanup.

cvializ committed:
🐛 Fix leap day dates. Add and unskip integration tests. (#20572)

* Only prevent users from entering non-leap-year 02/29 after 1000AD

* Add regression test for leap years. Factor out utility

* Use year string length and not raw year numeric value.

* Fix amp-inputmask in single-pass mode

* Don't create a new global

* Fix dependency check

* Rename patch number

* Update build

Enriqe committed:
Add i18n label to twitter tooltip label (#20468)

* add i18n label to twitter tooltip label

* register localization service in tests, use localizationId instead of string, change label value.

* period

* localized string id

danielrozenberg committed:
Add visual diff tests for <amp-selector> (#20628)

Commit:
Add visual test for amp-sidebar. (#20629)

andrewwatterson committed:
Added first few tests for amp-date-picker. (#20632)

cvializ committed:
Add first amp-accordion visual tests (#20631)

* Add first amp-accordion visual tests

* Remove unneeded code

Commit:
amp-script: Partial opacity pre-hydration (#20623)

* Remove non-SSR todomvc JS.

* Hook up onCreateWorker and onHydration.

* Set opacity to 0.7 until hydrated.

alanorozco committed:
♻️✨ [auto-lightbox] Honor explicit opt-out, refactor tests (#20596)

- Honor explicitly opted out by [data-amp-auto-lightbox-disable].
- Refactor tests for length and clarity.

zhangsu committed:
Introduce a secure way to embed sandboxed AMP (#20634)

* Introduce a secure way to embed sandboxed AMP

For an AMP document embedded inside a sandbox iframe, the origin of the
document would be "null", which defeats the purpose of an origin check.
An attacker could simply create a sandboxed, malicious iframe (therefore
having null origin), walk on the DOM frame tree to find a reference to
the viewer iframe (this is not constrained bu the same origin policy),
and then send postMessage() calls to the viewer frame and pass the
viewer's origin checks, if any.

The viewer could also check the source of the message to be a legitimate
AMP iframe window, but the attacker could bypass that by navigating the
legitimate AMP iframe window away to a malicious document. Recent
browsers have banned this kind of attack, but it's tricky to rely on it.

To prevent the above attack in a null origin AMP document, the viewer
should include this token in an init parameter, either in the `src` or
`name` attribute of the iframe, and then verify that this token is
included in all the messages sent from AMP to the viewer. The attacker
would not be able to steal this token under the same origin policy,
because the token is inside the viewer document at a different origin
and the attacker can't access it.

* Fix typo

* Fix check-type

AmpViewerMessage is defined as a @struct, which can't be modified after
its construction. I also can't use the square bracket notation to mutate
it. I also don't want to modify the struct definition to allow the
`messagingToken` field, since I want to conditionally set it on the
message depending on whether it's undefined (as opposed to having
something like `messagingToken: undefined` in the object).

erwinmombay committed:
Revert "temporarily disable saucelabs because of flakiness (#20566)" (#20574)

* Revert "temporarily disable saucelabs because of flakiness (#20566)"

This reverts commit 72c3cfecadf1039efe6e11ce7812f162cec7567a.

* disable chrome

* disable chrome for now in saucelabs

* dont skip chrome beta since tests flakes dont cause the build to fail

* comment out edge for now

* disable chrome beta too

* turn on chrome beta

zikas committed:
✨Enable conversion tracking for GTAG. (#20591)

* Enable conversion tracking for GTAG.

* Added hasGcl and hasDcl variables.

* Added hasGcl and hasDcl variables.

* Added hasGcl and hasDcl variables.

torch2424 committed:
✅ amp-recaptcha-input: Added support for integration test mocking in express server (#20540)

* Started the integration test and mess around with bootstrap frame

* Left a note for monday

* Made tons of progress on the integration test

* Got the text replace working with the mock file

Need to write recaptcha.mock.js

* TIL Integration test watching wont build your extension.

* Finished integration test, fixed precommit errors

* Splitting up the PR

* Extra newline

* Wildcard matched all recaptcha frames for dev server

* Made requested changes

* Made more requested changes

* FIxed precommit, and added comments

jeffkaufman committed:
initial draft amp-mraid implementation (#19531)

* initial draft amp-mraid implementation

* responding to reviewer comments:

* switched amp-mraid from a custom element to a host api
* minor things

* rework validation doc

* move MraidService to its own file, minor tweaks

* finish splitting out mraid service

* rework error handling

* document why impedance matching isn't needed

* not_supported -> unsupported

* make lint and type checker happy

cathyxz committed:
Make sure that infinite scroll actually ends (#20647)

cvializ committed:
Fix mask-output when combined with amp-list (#20648)

boyntoni committed:
Add contextual article matching to JW Player AMP component

alabiaga committed:
launch amp-list-viewport-resize to all (#20643)

Commit:
✨ amp-skimlinks - new "custom-redirect-domain" option. (#20547)

Creating new amp-skimlinks option "custom-redirect-domain" for publishers to change the redirect URL.

Currently, if users click on a link affiliated by Skimlinks, the user will navigate first to `https://go.skimresources.com` (the default redirect domain) before being redirected to the merchant URL. With the new custom-redirect-domain attribute, the publisher can now specify its own redirect domain name. (Typically a publisher domain CNAME pointing to go.skimresources.com).
E.g:
```html
<amp-skimlinks
   ...
   custom-redirect-domain="go.androidpolice.com">
```
This configuration will send the user to `http://go.androidpolice.com` instead of `https://go.skimresources.com`.

**Note**: `waypointUrl` has been removed from the e2e tests JSON config since "custom-redirect-domain" option can now be used instead.

cvializ committed:
Fix deleting from the middle of dates (#20656)

Includes backspace, delete, cut and highlight replace.

boyntoni committed:
Update withArgs

boyntoni committed:
Resolve linter issues

Commit:
Update to apply correct e-mail

danielrozenberg committed:
More visual diff tests for `<amp-selector>` (#20633)

Commit:
Skip flaky amp-list test. (#20665)

Commit:
amp-script: "Hello world" integration test (#20626)

* Add simple integration test.

* Wait for hydration and increase poll timeout.

* Upgrade worker-dom to 0.2.15.

* Use 0.2.17 now.

* Increase mocha timeout on Travis from 10s to 15s.

* Try increasing spec.timeout default instead.

* Skip single pass for now.

aghassemi committed:
✨ amp-sidebar: support autoscroll  (#20524)

gmajoulet committed:
Set a default state for the embedded components. (#20660)

gmajoulet committed:
Forward touch events to the viewer when #cap=swipe. (#20670)

alanorozco committed:
🐛♻️✅ Relax document type restrictions for auto-lightbox (#20669)

- Don't try to match LD+JSON schema URL from `mainEntityOfPage` or `url`. These fields are not present in common scenarios where Google SERP and others would parse otherwise.

- Match by Open Graph `og:type` when `article`, as it's present in almost every single document that may have malformed LD+JSON schema.

- Add a couple missing tests.

- Restructure for testing DOM-selection of type-matching nodes.

Commit:
Remove amp-script demo link (#20672)

helin24 committed:
✨Scroll extension: Add tracking of article reading progress (#20222)

* Track and send latest paragraph read

* Separate tracking, other fixes

* Change name to features

* Reverse change in presubmit-checks

* Encode snippet

* Force snippet to be a string

* Use viewport change instead of position observer

* Search all paragraphs, mark private methods, use viewport from services

* Debounce finding last paragraph, record index

* Add tests for read depth tracking

* Use sandbox and permit use of innerText

* Condition tracking paragraphs on whenReady

Commit:
Updated tests to use async/await, updated readme

Commit:
Add missing await

aghassemi committed:
amp-sidebar: revert validation rules for now (#20693)

pbakaus committed:
Remove TOC from A4A spec, as it is redundant when displayed on our site (#20649)

The TOC would display twice as of now – this removes the hard coded one.

cvializ committed:
Skip flaky inabox tests (#20682)

* Skip flaky inabox tests

* Skip firefox instead

alin04 committed:
Validator rollup (#20695)

* cl/231693942 Revision bump for #20544

* cl/231839357 MatchingDispatchKey returns all TagSpecs that match.

* cl/232015594 Revision bump for #20427

* cl/232308889 Validate transformed AMP via type identifier.

* cl/232353446 Update tests to reflect ordering of reorder head transformer

* cl/232354800 Revision bump for #20547

* cl/232395416 Do not count non-data URLs in style amp-custom for transformed AMP

* Fix lint errors.

alanorozco committed:
✅♻️ Refactor and test devdash (#20651)

- Restructure and extract `template.js` modules for semantic grouping.
- Fold `html.js` and `html-helpers.js` into one.
- Test exported HTML renderer outputs by DOM selection.
- Test several helpers.
- Test top-level renderer output to ensure all devdash pages are AMPHTML valid!

zhouyx committed:
Add visual diff test for amp-user-notification (#20635)

* add visual diff test for amp-user-notification

* address comments

newmuis committed:
♻️ Migrate the amp-story analytics to a service (#20041)

* Migrate the amp-story analytics to a service

* Get service from util

* Update to renamed file

* Remove duplicate import

* Wrap line

* export service for type visibility

* Fix type

* Update test

alanorozco committed:
✨Autofocus on `show()` (#20687)

## Breakdown

- Autofocuses optional `[autofocus]` element in the displayed tree via `show`/`toggle`.
- Forces synchronous display-focus chain on iOS so that programmatic focus is honored.

## Usage

```html
  <button on="tap: wrapper.show">
    Show
  </button>
  <div id="wrapper" hidden>
    <input type="text" autofocus>
  </div>
```

## etc.

Demo: <https://amp-autofocus.glitch.me/>

Closes #19949.

Commit:
Update validator tests

Commit:
Add back closing tags
